### PR TITLE
Fix NPE when accessing a field with a default value

### LINKED ISSUE #43359: [Bug]: Getting NPE when accessing a field with a default value which is generated from rest binding with the optional field
### Description

I am creating a record A using a rest binding with another record B. The record B and record A has the same record typed field, but in record B the field is optional and in record A that field has a default value. When accessing a member of this fiedl from record A, I am getting a NPE. But it worked in in Ballerina SwanLake Update 8(2201.8.x)

### Steps to Reproduce

Use the following code to reproduce the issue:
```bal
import ballerina/http;
import ballerina/io;

public type ConnectionConfig record {|
    http:BearerTokenConfig|http:OAuth2RefreshTokenGrantConfig auth;
    http:HttpVersion httpVersion = http:HTTP_2_0;
    http:ClientHttp1Settings http1Settings?;
    http:ClientHttp2Settings http2Settings?;
    decimal timeout = 60;
    string forwarded = "disable";
    http:PoolConfiguration poolConfig?;
    http:CacheConfig cache?;
    http:Compression compression = http:COMPRESSION_AUTO;
    http:CircuitBreakerConfig circuitBreaker?;
    http:RetryConfig retryConfig?;
    http:ResponseLimitConfigs responseLimits?;
    http:ClientSecureSocket secureSocket?;
    http:ProxyConfig proxy?;
    boolean validation = true;
|};

public function main() {
    ConnectionConfig connectionConfig = {
        auth: {
            refreshToken: "mockRefreshToken",
            clientId: "mockClientId",
            clientSecret: "mockClientSecret",
            refreshUrl: "mockRefreshUrl"
        }
    };
    http:ClientConfiguration config = {...connectionConfig};
    io:println(config.cache.enabled);
}
```

Run the file:
```console
$ bal run test.bal

Compiling source
        test.bal

Running executable

[2024-09-04 12:38:39,740] SEVERE {b7a.log.crash} - Cannot invoke "io.ballerina.runtime.internal.values.MapValueImpl.getUnboxedBooleanValue(io.ballerina.runtime.api.values.BString)" because "<local37>" is null 
java.lang.NullPointerException: Cannot invoke "io.ballerina.runtime.internal.values.MapValueImpl.getUnboxedBooleanValue(io.ballerina.runtime.api.values.BString)" because "<local37>" is null
        at test_service.main(test_service.bal:32)
        at $_init.$moduleExecute(.)
        at $_init.$lambda$$moduleExecute$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:320)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:287)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

### Affected Version(s)

Ballerina SwanLake Update9(2201.9.x) and Update10(2201.10.x)

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
$subject

Fixes #43359

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java [hunk 1]
```diff
@@ -421,36 +421,19 @@ public void visit(BLangRecordTypeNode recordTypeNode) {
...
         }
         List<String> fieldNames = getFieldNames(recordTypeNode.fields);
         BTypeSymbol typeSymbol = recordTypeNode.getBType().tsymbol;
         String typeName = recordTypeNode.symbol.name.value;
-        PackageID packageID = typeSymbol.pkgID;
         for (BLangType type : recordTypeNode.typeRefs) {
             BType bType = type.getBType();
-            if (packageID.equals(bType.tsymbol.pkgID)) {
-                continue;
-            }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java:453]: Given that we're removing this, I think we can improve `generateClosureForDefaultValues` to create new closures only if the original closure is defined in a separate module?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java:453]: I guess we can do this separately though.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/record/OpenRecordTypeInclusionTest.java [hunk 2]
```diff
@@ -193,7 +193,8 @@ public Object[] testFunctions() {
                 "testCyclicRecord",
                 "testOutOfOrderFieldOverridingFieldFromTypeInclusion",
                 "testCreatingRecordWithOverriddenFields",
-                "testDefaultValuesOfRecordFieldsWithTypeInclusion"
+                "testDefaultValuesOfRecordFieldsWithTypeInclusion",
+                "defaultValueFromInclusion"
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/record/OpenRecordTypeInclusionTest.java:197]: Change the name to start with `test`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 3]
```diff
@@ -6190,9 +6190,9 @@ private BLangRecordLiteral.BLangRecordKeyValueField createRecordKeyValueField(Lo
     }
 
     public void generateFieldsForUserUnspecifiedRecordFields(BLangRecordLiteral recordLiteral,
-                                                              List<RecordLiteralNode.RecordField> userSpecifiedFields) {
+                                                             List<RecordLiteralNode.RecordField> userSpecifiedFields) {
         BType type = Types.getImpliedType(recordLiteral.getBType());
-        if (type.getKind() != TypeKind.RECORD) {
+        if (type.getKind() != TypeKind.RECORD || isSpreadingAnOpenRecord(userSpecifiedFields)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6195]: Why do we need this check again?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -6202,6 +6202,22 @@ public void generateFieldsForUserUnspecifiedRecordFields(BLangRecordLiteral reco
...
+    private static boolean isSpreadingAnOpenRecord(List<RecordLiteralNode.RecordField> userSpecifiedFields) {
+        for (RecordLiteralNode.RecordField field : userSpecifiedFields) {
+            if (!(field instanceof BLangRecordLiteral.BLangRecordSpreadOperatorField spreadOperatorField)) {
+                continue;
+            }
+            BType type = Types.getReferredType(spreadOperatorField.expr.getBType());
+            if (!(type instanceof BRecordType recordType)) {
+                return true;
+            }
+            if (recordType.restFieldType != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6214]: Can still be a closed record if the rest type is not null, but is explicitly or effectively never.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/record/OpenRecordTypeInclusionTest.java (hunk 2 above)
Comment (heshanpadmasiri) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/record/OpenRecordTypeInclusionTest.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6195]: We call this method for any `BLangRecordLiteral` [here](https://github.com/ballerina-platform/ballerina-lang/blob/bc7e5e948b6dadc96ea5c781c751904e98f3f3fc/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java#L6088), so we need to determine if the compiler need to explicitly add fields to the RecordLiteral or not in this method.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java:453]: To clarify in the fallowing code,
```
type OuterXBase record {
    Inner inner = {foo: 10};
};

type OuterX record {|
    *OuterXBase; // <- default value must come from the inclusion
|};
```

In current implementation we create a lambda for creating an `Inner` value as a default closure for `OuterXBase` (let's call this `f1`), and for `OuterX` we create a lambda (let's call this `f2`) that simply invoke `f1`.  Are you suggesting in both cases they should have a single lambda (`f1`)

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -10119,7 +10136,7 @@ private void handleSafeNavigation(BLangBlockStmt blockStmt, BLangAccessExpressio
 
         boolean isAllTypesRecords = false;
         LinkedHashSet<BType> memTypes = new LinkedHashSet<>();
-        BType referredType = Types.getImpliedType(matchExpr.getBType());
+        BType referredType = getImpliedType(matchExpr.getBType());
```
Comment (SasinduDilshara) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:10139]: Do we need these changes in this PR

File: tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal [hunk 6]
```diff
@@ -289,9 +289,107 @@ function testDefaultValuesOfRecordFieldsWithTypeInclusion() {
...
+    OuterXAlsoClosed oxx = {...o};
+    assertEquality(oxx.inner.foo, 10);
+    lock {
+        assertEquality(4, count);
+    }
+    OuterXAlsoClosed oxx1 = {...o1};
+    assertEquality(20, oxx1.inner.foo);
+    lock {
+        assertEquality(4, count);
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal:387]: Can we also add a test or two where there are non-spread fields and more than one spread field?

File: tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal [hunk 7]
```diff
@@ -289,9 +289,107 @@ function testDefaultValuesOfRecordFieldsWithTypeInclusion() {
...
+}
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal:389]: Can we also add a simple test (in the appropriate files) for the rest scenario we've fixed?

Something along the lines of

```ballerina
import ballerina/io;

type Employee record {
    string id = fn();
    string name;
};

type OpenRecord record {|
    string name;
    string...;
|};

isolated function fn() returns string {
    io:println("fn");
    return "default";
}

public function main() {
    OpenRecord or = {name: "May", "id": "A1234"};
    Employee emp = {...or};
    io:println(emp);
}
```

File: tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal:387]: fixed

File: tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/open_record_type_inclusion.bal:421]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureGenerator.java:453]: Added a comment to  #41949

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6197]: Added comment explaining why we need this

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 8]
```diff
@@ -6202,6 +6205,23 @@ public void generateFieldsForUserUnspecifiedRecordFields(BLangRecordLiteral reco
...
+        for (RecordLiteralNode.RecordField field : userSpecifiedFields) {
+            if (!(field instanceof BLangRecordLiteral.BLangRecordSpreadOperatorField spreadOperatorField)) {
+                continue;
+            }
+            BType type = Types.getReferredType(spreadOperatorField.expr.getBType());
+            if (!(type instanceof BRecordType recordType)) {
+                return true;
+            }
+            if (recordType.restFieldType != null &&
+                    Types.getImpliedType(recordType.restFieldType).tag != TypeTags.NEVER) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6218]: The more accurate check here is org.wso2.ballerinalang.compiler.semantics.analyzer.Types#isNeverTypeOrStructureTypeWithARequiredNeverMember(org.wso2.ballerinalang.compiler.semantics.model.types.BType).

**Pull request merged**